### PR TITLE
(feat) O3-3036: Link the schema editor to the O3 standard JSON schema and validate schema

### PR DESCRIPTION
**Pull request by hadijahkyampeire**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary
- This links the schema editor to our [standard schema](https://json.openmrs.org/form.schema.json).
- It then validates the schema based on the standard JSON schema.
- It highlights errors in the Schema.
- It displays error details in a  carbon  `inline notification` and paginates them

## Recording

https://github.com/openmrs/openmrs-esm-form-builder/assets/30952856/0452372d-8ceb-454e-8ec8-4a9a04147446

## Related Issue
- https://openmrs.atlassian.net/browse/O3-3036

### CODE REVIEW

File: src/components/schema-editor/schema-editor.component.tsx [hunk 1]
```diff
@@ -2,49 +2,132 @@ import React from 'react';
...
   onSchemaChange: (stringifiedSchema: string) => void;
   stringifiedSchema: string;
+  errors: Array<markerProps>;
+  setErrors: (errors: Array<markerProps>) => void;
 }
 
-const SchemaEditor: React.FC<SchemaEditorProps> = ({ invalidJsonErrorMessage, onSchemaChange, stringifiedSchema }) => {
-  const { t } = useTranslation();
+const SchemaEditor: React.FC<SchemaEditorProps> = ({ onSchemaChange, stringifiedSchema, errors, setErrors }) => {
+  const fetchAndParseSchema = async () => {
```
Comment (arodidev) [inline src/components/schema-editor/schema-editor.component.tsx:31]: Could we move this to a resources folder, preferrably have it as a hook?

File: src/components/schema-editor/schema-editor.component.tsx [hunk 2]
```diff
@@ -2,49 +2,132 @@ import React from 'react';
...
+      setErrors(errorMarkers);
+    } else {
+      setErrors([]);
+    }
+  };
 
+  const handleChange = (newValue: string) => {
+    onSchemaChange(newValue);
+    fetchAndParseSchema()
+      .then((schema) => {
```
Comment (arodidev) [inline src/components/schema-editor/schema-editor.component.tsx:80]: In relation to the comment above, I think we should have this network request as a hook at the top of the component, making this fetch request on every change seems like an expensive process, also might be more efficient making this request using SWR.

File: src/components/schema-editor/schema-editor.component.tsx [hunk 3]
```diff
@@ -2,49 +2,132 @@ import React from 'react';
...
         showPrintMargin={false}
         showGutter={true}
         highlightActiveLine={true}
         value={stringifiedSchema}
         setOptions={{
-          enableBasicAutocompletion: false,
-          enableLiveAutocompletion: false,
+          enableBasicAutocompletion: true,
+          enableLiveAutocompletion: true,
           displayIndentGuides: true,
```
Comment (arodidev) [inline src/components/schema-editor/schema-editor.component.tsx:123]: I remember us testing the `enableBasicAutocompletion`, did you manage to get this working?

File: src/components/schema-editor/schema-editor.component.tsx [hunk 4]
```diff
@@ -2,49 +2,132 @@ import React from 'react';
...
         fontSize={15}
         showPrintMargin={false}
         showGutter={true}
         highlightActiveLine={true}
         value={stringifiedSchema}
         setOptions={{
-          enableBasicAutocompletion: false,
-          enableLiveAutocompletion: false,
+          enableBasicAutocompletion: true,
+          enableLiveAutocompletion: true,
```
Comment (arodidev) [inline src/components/schema-editor/schema-editor.component.tsx:122]: I remember us testing the enableBasicAutocompletion, did you manage to get this working?

**Review by arodidev — COMMENTED** [review]
Thanks for the great work @hadijahkyampeire, left a few review comments.

File: src/components/schema-editor/schema-editor.component.tsx (hunk 2 above)
Comment (hadijahkyampeire) [reply src/components/schema-editor/schema-editor.component.tsx:80]: Sure.

File: src/components/schema-editor/schema-editor.component.tsx (hunk 4 above)
Comment (hadijahkyampeire) [reply src/components/schema-editor/schema-editor.component.tsx:122]: No it did not work but that's basically what the documents say I do, turning it to true. I will look at it in a follow up PF,

**Comment by hadijahkyampeire** [comment]
@arodidev I have addressed the changes to move the JSON request in a hook, but for the autocomplete, I will handle it in another ticket.

File: src/hooks/useStandardFormSchema.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import useSWR from 'swr';
+import { openmrsFetch } from '@openmrs/esm-framework';
+
+export function useStandardFormSchema() {
+  const url = 'https://json.openmrs.org/form.schema.json';
+  const { data, error, isLoading } = useSWR<{ data }, Error>(url, openmrsFetch);
```
Comment (denniskigen) [inline src/hooks/useStandardFormSchema.ts:6]: Prefer [useSWRImmutable](https://swr.vercel.app/docs/revalidation#disable-automatic-revalidations) here as the schema is unlikely to be a resource that changes often.

File: src/components/schema-editor/schema-editor.component.tsx [hunk 6]
```diff
@@ -2,49 +2,142 @@ import React from 'react';
...
+      <ToggletipButton label="Show information">
+        <WarningFilled />
+      </ToggletipButton>
+      <ToggletipContent>
+        {errors.map((e) => (
+          <div>{e.text}</div>
+        ))}
+        <ToggletipActions>
+          <Link href="https://json.openmrs.org/form.schema.json" target="_blank">
+            Reference Schema
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:103]: Ditto.

File: src/components/schema-editor/schema-editor.component.tsx [hunk 7]
```diff
@@ -2,49 +2,142 @@ import React from 'react';
...
+  const debouncedValidateJSON = debounce(validateJSON, 300);
+
+  const handleChange = (newValue: string) => {
+    onSchemaChange(newValue);
+    debouncedValidateJSON(newValue, schema);
+  };
+
+  const ErrorMessages = () => (
+    <Toggletip className={styles.toggletipContainer} align="top-left">
+      <ToggletipButton label="Show information">
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:94]: This label should be run through the translation system.

File: src/components/schema-editor/schema-editor.component.tsx [hunk 8]
```diff
@@ -2,49 +2,142 @@ import React from 'react';
...
+        <WarningFilled />
+      </ToggletipButton>
+      <ToggletipContent>
+        {errors.map((e) => (
+          <div>{e.text}</div>
+        ))}
+        <ToggletipActions>
+          <Link href="https://json.openmrs.org/form.schema.json" target="_blank">
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:102]:
```suggestion
          <Link href="https://json.openmrs.org/form.schema.json" target="_blank" rel="noopener noreferrer">
```

File: src/components/schema-editor/schema-editor.component.tsx [hunk 9]
```diff
@@ -2,49 +2,142 @@ import React from 'react';
...
+          const instancePath = error.instancePath.replace(/^\./, '');
+          for (let i = 0; i < jsonLines.length; i++) {
+            if (jsonLines[i].includes(instancePath)) {
+              lineNumber = i;
+              break;
+            }
+          }
+          const message =
+            error.keyword === 'type'
+              ? `Invalid Type: ${error.instancePath.substring(1)} ${error.message}`
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:61]: Ditto

File: src/components/form-editor/form-editor.component.tsx [hunk 10]
```diff
@@ -306,12 +328,15 @@ const FormEditor: React.FC = () => {
...
+                        renderIcon={isMaximized ? Minimize : Maximize}
                         kind={'ghost'}
-                        iconDescription={t('downloadSchema', 'Download schema')}
+                        iconDescription={
+                          isMaximized ? t('minimizeEditor', 'Minimize editor') : t('maximizeEditor', 'Maximize editor')
+                        }
                         hasIconOnly
                         size="md"
                         tooltipAlignment="start"
+                        onClick={handleToggleMaximize}
```
Comment (denniskigen) [inline src/components/form-editor/form-editor.component.tsx:339]: What happens to the stuff related to downloading the sample schema?

File: src/components/form-editor/form-editor.component.tsx [hunk 11]
```diff
@@ -39,6 +39,17 @@ interface ErrorProps {
   title: string;
 }
 
+interface markerProps {
```
Comment (denniskigen) [inline src/components/form-editor/form-editor.component.tsx:42]:
```suggestion
interface MarkerProps {
```

File: src/components/form-editor/form-editor.component.tsx [hunk 12]
```diff
@@ -39,6 +39,17 @@ interface ErrorProps {
...
+  startCol: number;
+  startRow: number;
+  endCol: number;
+  endRow: number;
+  className: string;
+  // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents, @typescript-eslint/no-explicit-any
+  type: string | any;
```
Comment (denniskigen) [inline src/components/form-editor/form-editor.component.tsx:49]:
```suggestion
  type: string | unknown;
```

Does `unknown` work here?

**Review by denniskigen — APPROVED** [review]
Mostly LGTM. I've left a few suggestions.

File: src/components/form-editor/form-editor.component.tsx (hunk 12 above)
Comment (hadijahkyampeire) [reply src/components/form-editor/form-editor.component.tsx:49]: unknown doesn't work, I had used it at first but it was failing.

File: src/components/form-editor/form-editor.component.tsx (hunk 10 above)
Comment (hadijahkyampeire) [reply src/components/form-editor/form-editor.component.tsx:339]: @denniskigen what do you mean by this? I am sorry I don't understand the question.

File: src/components/schema-editor/schema-editor.component.tsx (hunk 9 above)
Comment (hadijahkyampeire) [reply src/components/schema-editor/schema-editor.component.tsx:61]: @denniskigen do you mean translating this one too?

File: src/components/schema-editor/schema-editor.component.tsx (hunk 7 above)
Comment (hadijahkyampeire) [reply src/components/schema-editor/schema-editor.component.tsx:94]: @denniskigen as if I remember Ian mentioning that carbon labels don't need to be translated, I don't think they will even be visible in the UI.

File: src/components/form-editor/form-editor.component.tsx (hunk 10 above)
Comment (hadijahkyampeire) [reply src/components/form-editor/form-editor.component.tsx:339]: This seems to already be on the main, I am not sure how the commit ended up in my commit.

**Comment by denniskigen** [comment]
Thoughts on this UI/UX?

CleanShot 2024-05-03 at 12  02 46@2x

**Comment by hadijahkyampeire** [comment]
@denniskigen I have updated the PR with the UI as suggested by Ciaran.

**Comment by arodidev** [comment]
LGTM @hadijahkyampeire

**Review by arodidev — APPROVED** [review]

**Comment by hadijahkyampeire** [comment]
@arodidev did you revert the approval?

**Comment by arodidev** [comment]
> @arodidev did you revert the approval?

No, I did not. Not sure how it got reverted.

**Comment by hadijahkyampeire** [comment]
@arodidev It says I reverted a stale review, not sure what that means.

**Comment by hadijahkyampeire** [comment]
@denniskigen I think this is now good to go in.

File: src/components/schema-editor/schema-editor.component.tsx [hunk 13]
```diff
@@ -84,21 +111,134 @@ const SchemaEditor: React.FC<SchemaEditorProps> = ({ invalidJsonErrorMessage, on
...
+  };
+
+  const onNextErrorClick = () => {
+    setCurrentIndex((prevIndex) => Math.min(prevIndex + 1, errors.length - 1));
+  };
+
+  const ErrorMessages = () =>
+    errors.length > 0 && (
+      <div className={styles.validationErrorsContainer}>
+        <ErrorNotification text={errors[currentIndex].text} line={errors[currentIndex].startRow} />
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:214]:
```suggestion
        <ErrorNotification text={errors[currentIndex]?.text} line={errors[currentIndex]?.startRow} />
```

Did we see on the call today that we might need to use optional chaining to access these values?

File: src/components/schema-editor/schema-editor.component.tsx [hunk 14]
```diff
@@ -84,21 +111,134 @@ const SchemaEditor: React.FC<SchemaEditorProps> = ({ invalidJsonErrorMessage, on
...
   }, [autocompleteSuggestions]);
 
-  return (
-    <>
-      {invalidJsonErrorMessage ? (
-        <div className={styles.errorContainer}>
-          <p className={styles.heading}>{t('schemaError', "There's an error in your schema.")}</p>
-          <p>{invalidJsonErrorMessage}</p>
+  // Validate JSON schema
+  const validateJSON = (content: string, schema) => {
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:115]:
```suggestion
  const validateSchema = (content: string, schema) => {
```

File: src/components/schema-editor/schema-editor.component.tsx [hunk 15]
```diff
@@ -84,21 +111,134 @@ const SchemaEditor: React.FC<SchemaEditorProps> = ({ invalidJsonErrorMessage, on
...
+        setErrors(errorMarkers);
+      } else {
+        setErrors([]);
+      }
+    } catch (error) {
+      console.log('Error during JSON parsing or validation:', error); // eslint-disable-line
+    }
+  };
+
+  const debouncedValidateJSON = debounce(validateJSON, 300);
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:181]:
```suggestion
  const debouncedValidateSchema = debounce(validateSchema, 300);
```

File: src/components/schema-editor/schema-editor.scss [hunk 16]
```diff
@@ -20,6 +20,33 @@
...
+
+  :global(.cds--actionable-notification) {
+    max-width: 38rem;
+  }
+  :global(.cds--actionable-notification__details) {
+    margin: 0 0rem 0 .8125rem !important;
+  }
+}
+
+
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.scss:35]:
```suggestion

```

File: src/components/schema-editor/schema-editor.scss [hunk 17]
```diff
@@ -20,6 +20,33 @@
...
+  gap: 0.5rem;
+  font-size: 0.875rem;
+  .paginationIcon {
+    fill: colors.$blue-50;
+    cursor: pointer;
+  }
+  .disabledIcon {
+    fill: colors.$gray-40;
+    cursor: not-allowed
+  }
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.scss:48]:
```suggestion
  display: flex;
  align-items: center;
  gap: 0.5rem;
  font-size: 0.875rem;
  
  .paginationIcon {
    fill: colors.$blue-50;
    cursor: pointer;
  }
  
  .disabledIcon {
    fill: colors.$gray-40;
    cursor: not-allowed;
  }
```

File: src/components/schema-editor/schema-editor.component.tsx [hunk 18]
```diff
@@ -84,21 +111,134 @@ const SchemaEditor: React.FC<SchemaEditorProps> = ({ invalidJsonErrorMessage, on
...
+  const ErrorNotification = ({ text, line }) => (
+    <ActionableNotification
+      subtitle={text}
+      inline={true}
+      title={`Error, line ${line + 1}:`}
+      kind="error"
+      lowContrast={true}
+      onActionButtonClick={() => window.open('https://json.openmrs.org/form.schema.json', '_blank')}
+      actionButtonLabel={t('referenceSchema', 'Reference schema')}
+    />
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:200]:
```suggestion
    <ActionableNotification
      inline
      lowContrast
      kind="error"
      subtitle={text}
      title={t('errorOnLine', 'Error on line') + ` ${line + 1}: `}
      actionButtonLabel={
        <Link
          target="_blank"
          rel="noopener noreferrer"
          href="https://json.openmrs.org/form.schema.json"
        >
          {t('referenceSchema', 'Reference schema')}
        </Link>
      }
    />
```

File: src/components/form-editor/form-editor.component.tsx [hunk 19]
```diff
@@ -40,6 +40,17 @@ interface ErrorProps {
...
+interface MarkerProps {
+  startCol: number;
+  startRow: number;
+  endCol: number;
+  endRow: number;
+  className: string;
+  // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents, @typescript-eslint/no-explicit-any
+  type: string | any;
+  text: string;
+}
```
Comment (denniskigen) [inline src/components/form-editor/form-editor.component.tsx:52]:
```suggestion
import { type IMarker } from 'react-ace';

// other code...

interface MarkerProps extends IMarker {
  text: string;
}

// And then later on in the `validateJson` function:

 return {
  startRow: lineNumber,
  startCol: 0,
  endRow: lineNumber,
  endCol: 1,
  className: 'error',
  text: message,
  type: 'text' as const,
};
```

**Review by denniskigen — COMMENTED** [review]
Thanks, @hadijahkyampeire. This is close. I've left some comments about where we could improve.

File: src/components/form-editor/form-editor.component.tsx (hunk 19 above)
Comment (hadijahkyampeire) [reply src/components/form-editor/form-editor.component.tsx:52]: @denniskigen this still complains about extending IMarker `Interface 'MarkerProps' incorrectly extends interface 'IMarker'.`

Comment (denniskigen) [reply src/components/form-editor/form-editor.component.tsx:52]: Mind showing the specific error?

File: src/components/schema-editor/schema-editor.scss [hunk 20]
```diff
@@ -20,6 +20,35 @@
...
+  font-size: 0.875rem;
+
+  .paginationIcon {
+    fill: colors.$blue-50;
+    cursor: pointer;
+  }
+  
+  .disabledIcon {
+    fill: colors.$gray-40;
+    cursor: not-allowed
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.scss:48]:
```suggestion
    cursor: not-allowed;
```

File: src/components/schema-editor/schema-editor.component.tsx [hunk 21]
```diff
@@ -84,21 +111,137 @@ const SchemaEditor: React.FC<SchemaEditorProps> = ({ invalidJsonErrorMessage, on
...
+  };
+
+  // Schema Validation Errors
+  const ErrorNotification = ({ text, line }) => (
+    <ActionableNotification
+      subtitle={text}
+      inline={true}
+      title={t('errorOnLine', 'Error on line') + ` ${line + 1}: `}
+      kind="error"
+      lowContrast={true}
```
Comment (denniskigen) [inline src/components/schema-editor/schema-editor.component.tsx:197]:
```suggestion
      inline
      title={t('errorOnLine', 'Error on line') + ` ${line + 1}: `}
      kind="error"
      lowContrast
```

**Comment by hadijahkyampeire** [comment]
@denniskigen This is the error I get when I try to extend `IMarker` and the markers prop in the AceEditor gets flagged as well
Screenshot 2024-05-09 at 08 26 27

File: src/components/form-editor/form-editor.component.tsx (hunk 19 above)
Comment (hadijahkyampeire) [reply src/components/form-editor/form-editor.component.tsx]: Thanks, @denniskigen, this is now fixed.

**Review by denniskigen — APPROVED** [review]
LGTM. Thanks, @hadijahkyampeire!

**Comment by hadijahkyampeire** [comment]
Thanks @denniskigen

**Comment by arodidev** [comment]
> @denniskigen This is the error I get when I try to extend `IMarker` and the markers prop in the AceEditor gets flagged as well Screenshot 2024-05-09 at 08 26 27

@hadijahkyampeire what would happen if we extended `type` with 'text' instead of 'string'?

**Pull request merged**